core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt

```diff
@@ -1,79 +1,39 @@
- */
-open class Capacitor : Port() {
-    override var name: String = "c"
+    override val resistance: Double
+      get() = if (circuit == null) { 0.0 } else { circuit!!.timeStep / capacitance }
 
-    /**
-     * The current, in Amperes, presently sourced by this Norton system.
-     */
-    internal var i: Double = 0.0
-        set(value) {
-            if (isInCircuit)
-                circuit!!.stampCurrentSource(pos.index, neg.index, value - field)
-            field = value
-        }
-
-    /**
-     * Capacitance, in Farads (Coulombs / Volt)
-     */
-    var capacitance: Double = 1e-5
-
-    /**
-     * The simulation timestep in seconds.
-     *
-     * This is set in [preStep], but the value is unfortunately not available during [stamp]; thus, it may be slightly out of date when [step] is actually called.
-     */
-    var ts: Double = 0.05  // A safe default
+    val charge : Double
+        get() = capacitance * potential
 
-    /**
-     * The "equivalent resistance" of the Norton system, in Ohms.
-     */
-    private val eqR: Double
-        get() = ts / capacitance
+    private var internalCurrentSource = CurrentSource()
 
-    /**
-     * The current, in Amperes, presently sourced by this Norton system.
-     */
-    internal var current: Double = 0.0
-        set(value) {
-            if (isInCircuit)
-                circuit!!.stampCurrentSource(pos.index, neg.index, value - field)
-            field = value
-        }
-
-    /**
-     * The current ideal potential across nodes, signed [pos] - [neg] in Volts, a function of the capacitance and the charge (integration of previous currents).
-     */
-    var idealU: Double = 0.0
+    override val current: Double
+        get() = if (resistance > 0.0) internalCurrentSource.current else 0.0
 
-    override fun detail(): String {
-        return "[capacitor c:$capacitance]"
+    override fun added() {
+        internalCurrentSource.circuit = circuit
+        internalCurrentSource.nodes.add(neg)
+        internalCurrentSource.nodes.add(pos)
     }
 
-    override fun preStep(dt: Double) {
-        ts = dt
-        current = -idealU / eqR
-        dprintln("C.preS: i=$current eqR=$eqR idealU=$idealU")
+    override fun removed() {
+        internalCurrentSource.circuit = null
+        internalCurrentSource.nodes.clear()
     }
 
-
-    override fun postStep(dt: Double) {
-        idealU = potential
-        dprintln("C.postS: potential=$potential (-> idealU)")
+    override fun stamp() {
+        if (resistance > 0.0) {
+            circuit!!.stampResistor(pos.index, neg.index, resistance)
+            internalCurrentSource.current = -potential / resistance
+            internalCurrentSource.stamp()
+        }
     }
 
-    override fun stamp() {
-        pos.stampResistor(neg, eqR)
+    override fun detail(): String {
+        return "$name: V = $potential, I = $current, R = $resistance, C = $capacitance, Q = $charge"
```

Comment:
I like the idea of adding more details about the components, but I did have a template such that the component is contained in `[]` brackets for clarity.

eg, `[capacitor $name: ${potential}v, ${current}A, ${resistance}Ω, ${capacitance}F, ${charge}J]` would be neat.